src/client/QXmppOmemoManager.cpp

```diff
@@ -0,0 +1,4998 @@
+
+            auto devicesCount = std::accumulate(recipientJids.cbegin(), recipientJids.cend(), 0, [=](const auto sum, const auto &jid) {
+                return sum + devices.value(jid).size();
+            });
+
+            // Do not exceed the maximum of manageable devices.
+            if (devicesCount > maximumDevicesPerStanza) {
+                q->warning(u"OMEMO payload could not be encrypted for all recipients because their devices are altogether more than the maximum of manageable devices " % QString::number(maximumDevicesPerStanza) % u" - Use QXmppOmemoManager::setMaximumDevicesPerStanza() to increase the maximum");
+                devicesCount = maximumDevicesPerStanza;
+            }
+
+            if (devicesCount) {
+                auto omemoElement = std::make_shared<QXmppOmemoElement>();
+                auto processedDevicesCount = std::make_shared<int>(0);
+                auto successfullyprocessedDevicesCount = std::make_shared<int>(0);
+
+                // Add envelopes for all devices of the recipients.
+                for (const auto &jid : recipientJids) {
+                    auto &recipientDevices = devices[jid];
+
+                    for (auto itr = recipientDevices.begin(); itr != recipientDevices.end(); ++itr) {
+                        const auto deviceId = itr.key();
+                        auto &device = itr.value();
+
+                        // Skip encrypting for a device if it does not respond
+                        // for a while.
+                        if (device.unrespondedSentStanzasCount == UNRESPONDED_STANZAS_UNTIL_ENCRYPTION_IS_STOPPED) {
+                            continue;
+                        }
+
+                        auto controlDeviceProcessing = [=](const bool isSuccessful) mutable {
+                            if (isSuccessful) {
+                                ++(*successfullyprocessedDevicesCount);
+                            }
+
+                            if (++(*processedDevicesCount) == devicesCount) {
+                                if (*successfullyprocessedDevicesCount == 0) {
+                                    q->warning("OMEMO element could not be created because no recipient devices with keys having accepted trust levels could be found");
+                                    reportFinishedResult(interface, {});
+                                } else {
+                                    omemoElement->setSenderDeviceId(ownDevice.id);
+                                    omemoElement->setPayload(payloadEncryptionResult.encryptedPayload);
+                                    reportFinishedResult(interface, { *omemoElement });
+                                }
+                            }
+                        };
+
+                        const auto address = Address(jid, deviceId);
+
+                        auto addOmemoEnvelope = [=, &device](const bool isKeyExchange) mutable {
```

Comment:
This should probably capture by value, as it could otherwise break if the jid is removed from the hash map.